core/modules/user/user.module

```diff
@@ -1094,34 +1094,32 @@ function user_register_access() {
 /**
  * User view access callback.
  *
- * @param $account
+ * @param User|AnonymousUser|object $account
  *   Can either be a full user object or a $uid.
+ *
+ * @return bool
+ *   Whether access is allowed or not.
  */
 function user_view_access($account) {
-  $uid = is_object($account) ? $account->uid : (int) $account;
-
-  // Never allow access to view the anonymous user account.
-  if ($uid) {
-    // Admins can view all, users can view own profiles at all times.
-    if ($GLOBALS['user']->uid == $uid || user_access('administer users')) {
-      return TRUE;
-    }
-    elseif (user_access('access user profiles')) {
-      // At this point, load the complete account object.
-      if (!is_object($account)) {
-        $account = user_load($uid);
-      }
-      return (is_object($account) && $account->status);
-    }
+  if (!($account instanceof User) && is_int($account)) {
+    $account = user_load($account);
   }
-  return FALSE;
+  return $account && $account->access('view');
 }
 
 /**
  * Access callback for user account editing.
  */
 function user_edit_access($account) {
-  return (($GLOBALS['user']->uid == $account->uid) || user_access('administer users')) && $account->uid > 0;
+  if (!($account instanceof User) && is_int($account)) {
+    $account = user_load($account);
+  }
```

Comment:
Looks like the old function did not allow $account to be an integer. Seems like we could nix this loading check. Note that menu access callbacks have their arguments "magic loaded" by the menu system (a `%user` portion of the path automatically calls `user_load()`), so this is always a full `User` object.

Reply:
user_edit_access is called directly by, for instance, views_handler_field_user_link_edit::render_link()

So it seems like it's not always a given that a proper user function is provided. That being said, the way the function was previously coded it did in deed expect a class.